### PR TITLE
Pass program name to kernel if possible

### PR DESCRIPTION
**Pull request by lmb**
From 4.15, a BPF program can have a name. This then shows up in debug output and other such nice things.

Pass this information to the kernel, if it is supported

### CODE REVIEW

File: prog.go [hunk 1]
```diff
@@ -42,34 +47,55 @@ type ProgramSpec struct {
...
-		return nil, fmt.Errorf("instructions cannot be empty")
+		return nil, errors.Errorf("instructions cannot be empty")
 	}
-	buf := bytes.NewBuffer(make([]byte, 0, len(spec.Instructions)*asm.InstructionSize))
-	err := spec.Instructions.Marshal(buf, nativeEndian)
+
+	detectHaveProgName.Do(func() {
+		attr, err := convertProgramSpec(&ProgramSpec{
+			Name: "feature_test",
+			Type: SocketFilter, // TODO: Socket Filter
```
Comment (arthurfabre) [inline prog.go]: old TODO?

File: prog.go [hunk 2]
```diff
@@ -42,34 +47,55 @@ type ProgramSpec struct {
...
 // NewProgram creates a new Program.
 func NewProgram(spec *ProgramSpec) (*Program, error) {
 	if len(spec.Instructions) == 0 {
-		return nil, fmt.Errorf("instructions cannot be empty")
+		return nil, errors.Errorf("instructions cannot be empty")
 	}
-	buf := bytes.NewBuffer(make([]byte, 0, len(spec.Instructions)*asm.InstructionSize))
-	err := spec.Instructions.Marshal(buf, nativeEndian)
+
+	detectHaveProgName.Do(func() {
```
Comment (arthurfabre) [inline prog.go]: Black magic! We should document that creating a program can cause more than one to be loaded :P

File: prog.go [hunk 3]
```diff
@@ -251,9 +309,44 @@ type loadError struct {
...
 func (le *loadError) Cause() error {
 	return le.cause
 }
+
+func convertCString(in []byte) string {
+	inLen := bytes.IndexByte(in, 0)
+	return string(in[:inLen])
+}
+
+func checkName(name string) error {
```
Comment (arthurfabre) [inline prog.go]: Would something like `regexp.MustCompile(fmt.Sprintf("^[0-9A-Za-z_]{0,%d}$", bpfObjNameLen))` be simpler?

Comment (lmb) [reply prog.go]: I guess it does, that should have occurred to me. This way we get better error messages though.

File: prog.go (hunk 2 above)
Comment (lmb) [reply prog.go]: Will do.

File: elf.go [hunk 4]
```diff
@@ -154,7 +154,13 @@ func (ec *elfCode) loadPrograms(progSections, relSections map[int]*elf.Section,
 			// labels they contain later on, and then link sections that way.
 			libs = append(libs, insns)
 		} else {
+			name := funcSym.Name
+			if len(name) > bpfObjNameLen-1 {
+				name = name[:bpfObjNameLen-1]
+			}
+
 			progs[funcSym.Name] = &ProgramSpec{
+				Name:          name,
```
Comment (arthurfabre) [inline elf.go]: What happens if there's a non `checkName()` character in the symbol name?

Comment (lmb) [reply elf.go]: It returns an error. Thankfully C functions can only contain those characters.

**Pull request merged**